rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,475 @@
+# RFC: Custom Styling
+
+## Contributors
+
+- Geoff Cox
+- Mason Tejera - MADS use cases
+- Ben Howell - easier recomposition
+- Micah Godbolt - component CSS vars
+
+## Summary
+
+Fluent UI React v9 is built today to adopt the Fluent 2 design as its default style. We strongly recommend that
+Microsoft applications leverage the default style and limit customization to a brand-based theme.
+
+Currently, v9 has mechanisms to support several customization scenarios:
+
+- Custom themes => Customize style values for all components
+- `classname` prop => Customize the style of one instance of a component
+- Extended design tokens => Support theme for a custom component or custom class/style
+- Hooks recomposition => Create a new component customizing the behavior, style, or rendering of an existing component.
+
+See the appendix for more detailed analysis of the capabilities and limitations of each of these mechansisms.
+
+## Problem
+
+Partners migrating from v0/v8 to v9 have brought up that v9's custom styling does not provide the same granularity of
+customization as previous versions. See the appendix for detailed partner use cases and scenarios.
+
+Partners producing component libaries or component design surfaces are the most concerned about the lack of
+customization. They would like to customize the style of a single type of component without changing the style of
+other components.
+
+Example: Change the border radius and border width of all Button components without changing those styles for
+Input components.
+
+## Option A: React.Context useCustomStyles hooks
+
+Each component could check context for a useCustomStyle hook within the useComponentStyle hook.
+If this hook was defined, it is called at the end of the method before returning state.
+
+> These example use Component as a substitute for a component like Button or Avatar.
+
+```tsx
+const customStyles = useContextSelector(ComponentCustomStylesContext, context => context?.useCustomStyles);
+
+//...component styling here
+
+return customStyles ? customStyles(state) : state;
+```
+
+Each component defines own types for props and state, so would need to define its own context value types,
```

Comment:
How do you expect nested overrides to work? Should the hook consume the outer context and call `customStyles` recursively?

Reply:
Ooo.  Hadn't thought of it from the point of view.  I was thinking it would only consume the closest containing context and wouldn't call any farther up the tree.

Reply:
If it will be a chain, it makes impossible to get rid of the previous custom styles:

```jsx
function App() {
  return (
    <ComponentCustomStylesContextProvider value={{ useCustomStyles }}>
      {/* I don't want this 👆 anymore 👿 */}
      <ComponentCustomStylesContextProvider value={{ useCustomStyles: null }}>
        {/* If this will be a chain, then oops 💥 */}
        <Component>Hello custom styles!</Component>
      </ComponentCustomStylesContextProvider>
    </ComponentCustomStylesContextProvider>
  );
}
```

Reply:
I think we need to think about the use cases and maybe there's a need for both? I think most cases people won't want to completely start from scratch but build off what is already there. That's what we have seen the most on our end and how the previous style model worked. However, I could also see it being useful to say "No, I want nothing that was here previously, wipe it all out". Could we pass an option to ignore context up the tree? This might also be a more performant path.

Reply:
The expectation is that customization would take place at the top level of the application.  In option A3, the object on context would be be replaced by another FluentProvider and we won't merge objects (lesson learned from v8).  Callers applying lower level customization will need to handle any merging of objects by spreading existing methods as the default to their customizations.